### PR TITLE
feat(httpClient): introducing an httpClient

### PR DESCRIPTION
**Pull request by Arqu**
This is the first step towards moving from RPC > HTTP.

So far this is non-breaking and simply introduces the HTTPClient and semi-defines the API endpoints we have.
Next series of PRs will convert us from RPC to the HTTPClient and funnel down the context from above to lib.

Also used the opportunity to update our ma/net package as the original one was deprecated and moved in as a subpackage.

### CODE REVIEW

File: lib/lib.go [hunk 1]
```diff
@@ -408,6 +408,19 @@ func NewInstance(ctx context.Context, repoPath string, opts ...Option) (qri *Ins
...
+			// TODO(arqu): I blindly assume this means we have a working API connection
+			// till the HTTPClient is fleshed out
+			maAddr, err := ma.NewMultiaddr(cfg.API.Address)
+			if err != nil {
+				return nil, err
+			}
+			httpAddr, err := manet.ToNetAddr(maAddr)
+			if err != nil {
+				return nil, err
+			}
```
Comment (b5) [inline lib/lib.go:421]: zomg I've wanted this function for a long time.

We should move these into the `NewHTTPClient` constructor though, and pass `cfg.API.Address` directly

File: lib/lib.go [hunk 2]
```diff
@@ -863,6 +877,14 @@ func (inst *Instance) RPC() *rpc.Client {
 	return inst.rpc
 }
 
+// HTTP accesses the instance HTTP client if one exists
+func (inst *Instance) HTTP() *HTTPClient {
```
Comment (b5) [inline lib/lib.go:881]: Any reason not to name this method `HTTPClient`?

File: config/api.go [hunk 3]
```diff
@@ -3,6 +3,7 @@ package config
 import (
```
Comment (b5) [inline config/api.go:3]: Not a fan of having all this defined in config. Can we move it to a new file `/lib/api.go`?

File: lib/http.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	// TODO(arqu): inject context values into headers
+
+	return c.do(ctx, addr, httpMethod, mimeType, params, result)
+}
+
+func (c HTTPClient) do(ctx context.Context, addr string, httpMethod string, mimeType string, params interface{}, result interface{}) error {
+	var req *http.Request
+	var err error
+
+	if httpMethod == "GET" || httpMethod == "DELETE" {
```
Comment (b5) [inline lib/http.go:48]:
```go
	if httpMethod == http.MethodGet || httpMethod == http.MethodDelete {
```

File: lib/http.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+// NewHTTPClient instantiates a new HTTPClient
+func NewHTTPClient(addr string) *HTTPClient {
+	return &HTTPClient{
+		Addr: addr,
+	}
+}
+
+// Call resolves the action by mapping it to the correct API endpoint and passes on parameters and context info
+// currently this is a fairly flat process but should evolve in the future
+func (c HTTPClient) Call(ctx context.Context, apiEndpoint config.APIEndpoint, methodOverride string, params interface{}, result interface{}) error {
```
Comment (b5) [inline lib/http.go:30]: Change `Call`'s signature to:

```go
func (c HTTPClient) Call(ctx context.Context, apiEndpoint config.APIEndpoint, params interface{}, result interface{}) error {
```

and have it proxy a second exported method `CallMethod`

```
func (c HTTPClient) CallMethod(ctx context.Context, apiEndpoint config.APIEndpoint, params interface{}, result interface{}) error {
```

passing in a default `http.MethodPost` when using `Call`

File: lib/http.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+// NewHTTPClient instantiates a new HTTPClient
+func NewHTTPClient(addr string) *HTTPClient {
+	return &HTTPClient{
+		Addr: addr,
+	}
+}
+
+// Call resolves the action by mapping it to the correct API endpoint and passes on parameters and context info
+// currently this is a fairly flat process but should evolve in the future
```
Comment (b5) [inline lib/http.go:29]: Comments like `currently this is...` get stale pretty quickly. I'd prefer we not add 'em

File: lib/http.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+func (c HTTPClient) Call(ctx context.Context, apiEndpoint config.APIEndpoint, methodOverride string, params interface{}, result interface{}) error {
+	// post should be the default method as we are almost always passing on a call from one instance to another
+	httpMethod := http.MethodPost
+	if methodOverride != "" {
+		httpMethod = methodOverride
+	}
+	// TODO(arqu): work out mimeType configuration/override per API endpoint
+	mimeType := jsonMimeType
+	addr := fmt.Sprintf("http://%s/%s", c.Addr, apiEndpoint)
```
Comment (b5) [inline lib/http.go:38]: we don't always use `HTTP`. I think this means the `NewHTTPClient` constructor is going to need to be more robust.

**Comment by dustmop** [comment]
No comments aside from b5's feedback.

**Review by b5 — APPROVED** [review]
Looks great! v excited about this one

**Pull request merged**